internal/command/meta_backend.go

```diff
@@ -721,6 +706,33 @@ func (m *Meta) backendFromConfig(opts *BackendOpts) (backend.Backend, tfdiags.Di
 	}
 }
 
+func (m *Meta) determineInitReason(previousBackendType string, currentBackendType string) string {
+	hasBackendTypeChanged := previousBackendType != currentBackendType
+	hasConfigBlockChanged := previousBackendType == currentBackendType
+
+	initReason := ""
+
+	if hasBackendTypeChanged {
+		if currentBackendType == "cloud" {
+			initReason = fmt.Sprintf("Backend configuration changed from %q to Terraform Cloud", previousBackendType)
+		} else if previousBackendType == "cloud" {
+			initReason = fmt.Sprintf("Backend configuration changed from Terraform Cloud to %q", currentBackendType)
+		} else {
+			initReason = fmt.Sprintf("Backend configuration changed from %q to %q", previousBackendType, currentBackendType)
+		}
+	}
+
+	if hasConfigBlockChanged {
+		if currentBackendType == "cloud" {
+			initReason = "Terraform Cloud configuration block changed"
+		} else {
+			initReason = "Backend configuration block changed"
+		}
+	}
+
+	return initReason
```

Comment:
While correct, I'd find this function easier to read if we had multiple `return` statements, rather than assigning a value to `initReason`. At first glance I thought the `if hasConfigBlockChanged` branch could overwrite the value of the `hasBackendTypeChanged` branch, but they're mutually exclusive.

Reply:
Agreed! Having multiple return statements rather than reassigning values to initReason would make this function more readable. I was going to make this suggested change when I notice that the PR that Chris mentioned I needed to wait for merging mine, had done a refactoring to same area of code I was refactoring: https://github.com/hashicorp/terraform/blob/310343fc1f4bda0f08f6b533961b33a39f8b876e/internal/command/meta_backend.go#L673-L713
I think Martin's refactoring looks great! So, at this point, I think there is very little refactoring I can add on top of this. I am just moving Martin's logic into its on function. Moving blocks of logic outside of the very long function `backendFromConfig()` may help with readability.
I am also reworded two of the strings assign to `initReason` because I recall Chris mentioning he wanted the word "block" as in "configuration block" in there.